primitives/src/json.rs

```diff
@@ -0,0 +1,201 @@
+//! String containing serialized valid JSON.
+//! This string is guaranteed to parse as JSON
+
+#[cfg(not(feature = "std"))]
+use alloc::{
+    borrow::ToOwned,
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+#[cfg(feature = "std")]
+use std::{
+    string::{String, ToString},
+    vec::Vec,
+};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
+
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Eq, Encode, Decode)]
+pub struct JsonString(pub String);
+
+/// A candidate for a valid `JsonString`.
+/// Is used for generalizing ser/de any types to `JsonString` and vise versa
+#[derive(Serialize, Deserialize, Clone)]
+struct JsonCandidate<T>(T);
+
+impl JsonString {
+    /// Construct [`JsonString`]
+    pub fn new<T: Serialize>(payload: T) -> Self {
+        JsonCandidate(payload).try_into().unwrap()
+    }
+
+    /// Tries cast [`JsonString`] to any value.
+    ///
+    /// # Errors
+    /// - if String == invalid Json type
+    pub fn try_into_any<T: DeserializeOwned>(&self) -> Result<T, serde_json::Error> {
+        serde_json::from_str(&self.0)
+    }
+}
+impl PartialEq for JsonString {
+    fn eq(&self, other: &Self) -> bool {
+        serde_json::from_str::<Value>(&self.0).unwrap()
+            == serde_json::from_str::<Value>(&other.0).unwrap()
+    }
+}
```

Comment:
IMO you can derive this. I think there is no point of having a special `PartialEq` if it can't be for all scenarios. The scenarios for which it fundamentally can't be implemented is `rose##alice@wonderland` vs `rose#wonderland@alice#wonderland`. How about you derive it and we open an issue to discuss whether we want to prevent. Whatever you decide, please open an issue so we can discuss the implementation of `Eq` and `Ord` for `JsonString`

Reply:
the problem there was that `Ord` implementation was not consistent with `Eq` implementation. They can compare just as strings, but they have to both do so